ENCM-115 Add functional characterization matrix

forresttanaka opened the pull request:
* The main addition exists in the matrix_functional_characterization.js file which mostly follows the pattern of the simpler matrix-rendering files.
* I changed `<MatrixAddCart>` to only request the `@id` and `@type` of the related datasets to reduce the size of the results, in some cases very radically. Those two properties are the only ones that the cart code uses.
* I changed `<SearchFilter>` to allow matrix pages to optionally pass the specific title they want for the search-box label.

forresttanaka committed:
ENCM-115-fc-matrix

* Now getting JSON for functional characterization matrix with both FunctionalCharacterizationExperiment and FunctionalCharacterizationSeries types without error. Getting 11 results while logged out locally.
* Initial embryonic implementation of the functional characterization matrix.
* Add facets to the page. Change the matrix row heading styling to fix its width.
* Make FC matrix menu URL closer to the proper one, but keep it partial for now to get more local data. Fix the Javascript crash when going to the URL with multiple type= query string elements. This also fixes the search-box label.
* Add a mouse-based functional characterization experiment.
* Non-functional commit just so I can get a demo running with the new FC matrix row configuration.
* Rewrite FC matrix generation completely.
* Correct column index for row data.
* Style the scientific name rows properly.
* Add a blank line between the species sections of the matrix.
* Display header with assay and perturbation type on separate lines.
* Correct the column index calculation that ignored the second perturbation value if an entry had both perturbed and non-perturbed entries.
* Don’t deduplicate the column keys when generating them, as we won’t find any duplicates anyway.
* Make header row links work correctly.
* Make links for the row subcategory labels.
* Change dumb spelling of the `no_perturbation` value when perturbation_type values are missing.
* Add links to the table cells.
* Make column header taller to fit long labels.
* First attempt at sorting the columns.
* Adjust height and font weight of column label cells.
* Update the column sorting according to what Ben told me in our 1×1.
* The column link query strings now only select species that exist in the current matrix data.
* Make mouse green less saturated.
* Update the URLs to their final form.
* Sort the species sections so that Homo sapiens always comes first.
* Correct the sorting from the last commit. I had it sorting the wrong thing.
* Add links for the rows that let you select a single species. Reduce the size of requests for adding matrix items to the cart.
* Use the context `search_base` to generate most links in the matrix.
* Change the FC matrix title to sentence case.
* Fix one of the BDD test failures now that the “Functional characterization” button no longer goes to a search-list page.
* Point at the renamed snovault-search branch.
* Correct a BDD test.
* Correct the BDD test button title in the last commit.
* Add BDD teset for the functional characterization matrix. Correct the home-page test link text again.
* Remove non-sensical functional characterization BDD test fragment.
* Update cart test for new experiments in the test data.
* Update cart test for new data inserts. Update comments.
* Update comments and remove unused code.

forresttanaka commented:
@emro Ugh. I left an old URL in the menu. It should be identical to the URL in the FC home-page button. I’ll push an update to that momentarily. That’s the only change.

forresttanaka committed:
ENCM-115-fc-matrix

Update FC Data menu query string.

emro reviewed (dismissed):
Looks awesome! I like the colors and it works well.

emro commented on src/encoded/search_views.py:
```diff
@@ -38,6 +38,7 @@
 from snosearch.fields import AllResponseField
 from snosearch.fields import BasicMatrixWithFacetsResponseField
 from snosearch.fields import MissingMatrixWithFacetsResponseField
+from snosearch.fields import MultitypeMissingMatrixWithFacetsResponseField
```
This is almost not worth noting but I might have done the cases like this: MultiTypeMissingMatrixWithFacetsResponseField

forresttanaka committed:
ENCM-115-fc-matrix

Try using `assay_title` instead of `assay_term_name` for the X-axis of the FC matrix so that MPRA and Control MPRA stay separate.

emro reviewed (approved):
Looks great!
